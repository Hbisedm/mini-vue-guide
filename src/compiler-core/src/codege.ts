export function generate(ast) {
  const context = createCodegenContent();
  const { push } = context;
  push("return ");

  push("const { toDisplayString: _toDisplayString } = Vue");

  push("return");

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(",");

  push(`function ${functionName}(${signature}){`);
  genNode(ast.codegenNode, context);
  push("}");

  return {
    code: context.code,
  };
}

function createCodegenContent() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
  };
  return context;
}

function genNode(codegenNode: any, context) {
  const { push } = context;
  const node = codegenNode;
  push(`return '${node.content}'`);
}
