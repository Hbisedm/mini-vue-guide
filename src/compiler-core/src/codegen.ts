import { isString } from "../../shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  helperMapName,
  TO_DISPLAY_STRING,
} from "./runtimeHelpers";

export function generate(ast) {
  const context = createCodegenContext();
  const { push } = context;

  getFunctionPreamble(ast, context);

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(",");

  push(`function ${functionName}(${signature}){`);
  push("return ");
  genNode(ast.codegenNode, context);
  push("}");

  return {
    code: context.code,
  };
}
/**
 * 处理前导码
 * @param ast
 * @param context
 */
function getFunctionPreamble(ast: any, context: any) {
  const { push } = context;
  const VueBinding = "Vue";
  // const helpers = ["toDisplayString"];
  const aliasHelper = (s) => `${helperMapName[s]}: _${helperMapName[s]}`;
  if (ast.helpers.length > 0) {
    push(
      `const { ${ast.helpers.map(aliasHelper).join(", ")} } from ${VueBinding}`
    );
  }
  push("\n");
  push("return ");
}

function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  };
  return context;
}

function genNode(node: any, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      getText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      gedInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
      break;
    default:
      break;
  }
}

function genCompoundExpression(node, context) {
  const { push } = context;
  const { children } = node;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(child, context);
    }
  }
}

function genElement(node: any, context: any) {
  const { push, helper } = context;
  const { tag, children, props } = node;
  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  // for (let i = 0; i < children.length; i++) {
  //   const child = children[i];
  //   genNode(child, context);
  // }
  // =>
  // const child = children[0];
  genNodeList(genNullable([tag, props, children]), context);
  push(")");
}

function genNodeList(nodes, context) {
  // Implement
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }
    if (i < nodes.length - 1) {
      push(", ");
    }
  }
}

function genNullable(args: any) {
  // Implement
  return args.map((arg) => arg || "null");
}

function genExpression(node, context) {
  // Implement
  const { push } = context;
  console.log(node);
  push(`${node.content}`);
}

function gedInterpolation(node, context: any) {
  /**
   * 插值 {{ 表达式类型的content }}
   */
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  // 处理表达式类型 ,使用统一的入口
  genNode(node.content, context);
  push(")");
}

function getText(node, context: any) {
  const { push } = context;
  push(`'${node.content}'`);
}
