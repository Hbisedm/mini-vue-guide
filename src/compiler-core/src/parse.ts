import { NodeTypes } from "./ast";

const enum TagType {
  START,
  END,
}

/**
 * content 统一的入口
 * @param content
 * @returns ast
 */
export function baseParse(content: string): any {
  const context = createParseContext(content);
  return createRoot(parseChildren(context, []));
}

function parseChildren(context, ancestors) {
  const nodes: any = [];
  while (!isEnd(context, ancestors)) {
    let node;
    const s = context.source;
    /**
     * {{}}
     * 处理插值表达式
     */
    if (s.startsWith("{{")) {
      node = parseInterpolation(context);
    } else if (s[0] === "<") {
      /**
       * 处理Element
       * 使用正则匹配 `<`开头 `[a-z]*`
       */
      if (/[a-z]/.test(s[1])) {
        node = parseElement(context, ancestors);
      }
    } else {
      node = parseText(context);
    }
    nodes.push(node);
  }
  return nodes;
}

function isEnd(context, ancestors) {
  const s = context.source;
  // Implement
  // 1. context.source 空
  // 2. tag
  if (s.startsWith("</")) {
    /** 优化倒序，提高遍历性能 */
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        // if (s.slice(2, 2 + tag.length) === tag) {
        return true;
      }
    }
  }

  return !s;
}

function parseText(context) {
  let endIndex = context.source.length;

  const endTokens = ["{{", "<"];

  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i]);
    /** 取出最小值 */
    if (index !== -1 && endIndex > index) {
      endIndex = index;
    }
  }
  let content = parseTextData(context, endIndex);
  // console.log("parseTexted ----");
  // console.log(content);

  return {
    type: NodeTypes.TEXT,
    content,
  };
}

function parseTextData(context: any, length) {
  /** 1. 获取值 */
  let content = context.source.slice(0, length);
  /** 2. 推进 */
  advanceBy(context, length);
  return content;
}

function parseElement(context, ancestors) {
  const element: any = parseTag(context, TagType.START);
  ancestors.push(element);
  element.children = parseChildren(context, ancestors);
  ancestors.pop();
  // console.log(" hbisedm Log...");
  // console.log(element.tag);
  // console.log(context.source);

  if (startsWithEndTagOpen(context.source, element.tag)) {
    // if (context.source.slice(2, 2 + element.tag.length) === element.tag) {
    parseTag(context, TagType.END);
  } else {
    throw new Error(`应该来个${element.tag}结尾`);
  }
  return element;
}

function startsWithEndTagOpen(source, tag) {
  return (
    source.startsWith("</") &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  );
}

function parseTag(context: any, tagType: TagType) {
  /**
   * 1. 解析tag
   */
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  const tag = match[1];

  /**
   * 2. 删除已处理完成的代码
   */
  advanceBy(context, match[0].length);
  advanceBy(context, 1);

  if (tagType === TagType.END) return;
  return {
    type: NodeTypes.ELEMENT,
    tag,
  };
}

function parseInterpolation(context) {
  // {{message}}
  const openDelimiter = "{{";
  const closeDelimiter = "}}";
  /**
   * 拿到字符尾部索引
   */
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );
  /**
   *  推进
   *  {{msg}}{{name}} => msg}}{{name}}
   */
  advanceBy(context, openDelimiter.length);

  /**
   * 计算出插值模版里面字符的长度
   */
  const rawContentLength = closeIndex - openDelimiter.length;

  const rawContent = parseTextData(context, rawContentLength);
  const content = rawContent.trim();
  /**
   * 继续推进
   * msg}}{{name}} => {{name}}
   */
  advanceBy(context, closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
}

function advanceBy(context, length) {
  context.source = context.source.slice(length);
}

/**
 * 封装成ctx对象
 * @param content
 * @returns
 */
function createParseContext(content: string): any {
  return {
    source: content,
  };
}

function createRoot(children) {
  return {
    children,
    type: NodeTypes.ROOT,
  };
}
