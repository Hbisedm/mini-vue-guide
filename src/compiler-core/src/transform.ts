import { TO_DISPLAY_STRING } from "./runtimeHelpers";
import { NodeTypes } from "./ast";

export function transform(root, options = {}) {
  const context = createContext(root, options);
  // 深度遍历
  traverseNodes(root, context);
  // 方便生成代码阶段调用转换后的ast树
  genCode(root);

  root.helpers = [...context.helpers.keys()];
}

function genCode(root) {
  root.codegenNode = root.children[0];
}

function traverseNodes(node: any, context) {
  const { nodeTransforms } = context;
  // 对节点进行用户自义定插件处理
  nodeTransforms.forEach((transform) => {
    transform(node, context);
  });
  // 处理不同类型
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
      break;
    default:
      break;
  }
}
function traverseChildren(root: any, context: any) {
  const children = root.children;
  for (let i = 0; i < children.length; i++) {
    const node = children[i];
    traverseNodes(node, context);
  }
}

function createContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
  };
  return context;
}
