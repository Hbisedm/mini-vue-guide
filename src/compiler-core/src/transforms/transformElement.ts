import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";
import { createVNodeCall, NodeTypes } from "../ast";

export function transformElement(node, context) {
  return () => {
    if (node.type === NodeTypes.ELEMENT) {
      //中间处理层

      // tag
      const vnodeTag = `"${node.tag}"`;

      // props
      let vnodeProps;

      const { children } = node;
      const vnodeChildren = children[0];

      node.codegenNode = createVNodeCall(
        context,
        vnodeTag,
        vnodeProps,
        vnodeChildren
      );
    }
  };
}
