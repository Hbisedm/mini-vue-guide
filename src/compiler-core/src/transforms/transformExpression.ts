import { NodeTypes } from "../ast";

/**
 * 专用处理插值的插件
 * @param node
 */
export function transformExpression(node) {
  if (node.type === NodeTypes.INTERPOLATION) {
    processExpression(node.content);
  }
}

function processExpression(node) {
  node.content = `_ctx.${node.content}`;
  return node;
}
