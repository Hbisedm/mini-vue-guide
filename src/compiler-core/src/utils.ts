import { NodeTypes } from "./ast";

/**
 * 判断当前节点是不是Text类型or插值类型
 * @param node
 * @returns
 */
export function isText(node) {
  return node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION;
}
