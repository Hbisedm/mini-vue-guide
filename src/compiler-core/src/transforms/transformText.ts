import { NodeTypes } from "../ast";
import { isText } from "../utils";

export function transformText(node) {
  return () => {
    if (node.type === NodeTypes.ELEMENT) {
      let currentContainer;
      const { children } = node;
      for (let i = 0; i < children.length; i++) {
        const child = children[i];
        // 如果是的话， 进行 `+`的操作
        if (isText(child)) {
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j];
            if (isText(next)) {
              // init
              if (!currentContainer) {
                // 更新 Text or插值为复合类型
                currentContainer = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [child],
                };
              }
              currentContainer.children.push(" + ");
              currentContainer.children.push(next);
              children.splice(j, 1);
              j--;
            } else {
              currentContainer = undefined;
              break;
            }
          }
        }
      }
    }
  };
}
