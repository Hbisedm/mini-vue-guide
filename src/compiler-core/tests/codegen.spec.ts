import { generate } from "../src/codege";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformElement } from "../src/transforms/transfromElement";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);

    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("element", () => {
    const ast = baseParse("<div></div>");
    transform(ast, {
      nodeTransforms: [transformElement],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
