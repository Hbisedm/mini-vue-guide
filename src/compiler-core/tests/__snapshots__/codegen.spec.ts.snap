// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { createElementVNode: _createElementVNode } from Vue
return function render(_ctx,_cache){return _createElementVNode('div')}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } from Vue
return function render(_ctx,_cache){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx,_cache){return 'hi'}"
`;
